<h1><center>Data Science Trainer Assignment</center></h1>
<h2><center>Future Skills</center></h2>




> Bindu Sri Vuyyuru

> bindusrivuyyuru0702@gmail.com





## Read data from google sheets

In [10]:
import pandas as pd
inpt1 = 1473172222 # '(Input)UserIDs'
inpt2 = 99428384 # '(Input)RigorBuilderRAW'
otpt1 = 321394201 # 'LeaderboardTeamWise(Output)'
otpt2 = 975448155 # 'LeaderboardIndividual(Output)'

In [11]:
User_Ids = pd.read_csv('https://docs.google.com/spreadsheets/d/11J1ivZqPOpIHs0_Ktfr8CDgXvik0PV80hUgRUCsCxvw/export?gid=' 
                       + str(inpt1)+ '&format=csv',header = 10,usecols = ['S No','Name','Team Name','User ID'],nrows= 21)
User_Ids

S No           Name            Team Name  User ID
0      1      Soumita M  Winning Culture Lab      280
1      2     Subhangi 0  Winning Culture Lab       75
2      3    Nitin Shane  Winning Culture Lab     1124
3      4         Merwin               Mentor      295
4      5  fardinkamal62    Human Capital Lab      263
5      6  __riddhi_213_    Human Capital Lab      504
6      7    Rohit Dutta               Mentor      266
7      8       imshawan             Tech Lab       71
8      9   Anuraj_Saini        BrandTech Lab      342
9     10        sharath  Student Unicorn Lab     3367
10    11        Ronak 0             Tech Lab      302
11    12  Amrit Malviya        Brandtech Lab      336
12    13        Saurabh               Mentor      271
13    14    darshimalde  Student Unicorn Lab     3169
14    15         Shagun             Tech Lab      100
15    16         Ayisha  Student Unicorn Lab     3406
16    17         Palash            Growpital       69
17    18          raman            Growpital      539
18    19        Nishant            Growpital      299
19    20         Vatsal              Kringle     3408
20    21       devmenkr              Kringle      360

In [12]:
rigor = pd.read_csv('https://docs.google.com/spreadsheets/d/11J1ivZqPOpIHs0_Ktfr8CDgXvik0PV80hUgRUCsCxvw/export?gid=' 
              + str(inpt2)+ '&format=csv',header = 7,usecols = ['S No','name','uid','total_statements','total_reasons'],nrows= 21)
rigor

S No           name   uid  total_statements  total_reasons
0      1      Soumita M   280                13             21
1      2     Subhangi 0    75                13             16
2      3    Nitin Shane  1124                12             11
3      4         Merwin   295                13             12
4      5  fardinkamal62   263                 2              2
5      6  __riddhi_213_   504                 2              7
6      7    Rohit Dutta   266                 3              3
7      8       imshawan    71                 9              9
8      9   Anuraj_Saini   342                 7              7
9     10        sharath  3367                 8              8
10    11        Ronak 0   302                16             19
11    12  Amrit Malviya   336                 2              4
12    13        Saurabh   271                 6              8
13    14    darshimalde  3169                 3              3
14    15         Shagun   100                 2              2
15    16         Ayisha  3406                11              8
16    17         Palash    69                 6              7
17    18          raman   539                 5              5
18    19        Nishant   299                 4              5
19    20         Vatsal  3408                 2              2
20    21       devmenkr   360                 7              7

In [13]:
team_leader_board = pd.read_csv('https://docs.google.com/spreadsheets/d/11J1ivZqPOpIHs0_Ktfr8CDgXvik0PV80hUgRUCsCxvw/export?gid=' 
              + str(otpt1)+ '&format=csv',header = 8,usecols = ['Team Rank','Thinking Teams Leaderboard','Average Statements','Average Reasons'],nrows= 9)
team_leader_board

Team Rank Thinking Teams Leaderboard  Average Statements  Average Reasons
0          1        Winning Culture Lab               12.67            16.00
1          2                   Tech Lab                9.00             9.33
2          3                     Mentor                7.33             7.67
3          4        Student Unicorn Lab                7.33             6.33
4          5                  Growpital                5.00             5.67
5          6              BrandTech Lab                4.50             5.50
6          7                    Kringle                4.50             4.50
7          8          Human Capital Lab                2.00             4.50

In [14]:
Individual_leader_board = pd.read_csv('https://docs.google.com/spreadsheets/d/11J1ivZqPOpIHs0_Ktfr8CDgXvik0PV80hUgRUCsCxvw/export?gid=' 
              + str(otpt2)+ '&format=csv',header = 7,usecols = ['Rank','Name','UID','No. of Statements','No. of Reasons'],nrows= 21)
Individual_leader_board

Rank           Name   UID  No. of Statements  No. of Reasons
0      1        Ronak 0   302                 16              19
1      2      Soumita M   280                 13              21
2      3     Subhangi 0    75                 13              16
3      4         Merwin   295                 13              12
4      5    Nitin Shane  1124                 12              11
5      6         Ayisha  3406                 11               8
6      7       imshawan    71                  9               9
7      8        sharath  3367                  8               8
8      9   Anuraj_Saini   342                  7               7
9     10       devmenkr   360                  7               7
10    11        Saurabh   271                  6               8
11    12         Palash    69                  6               7
12    13          raman   539                  5               5
13    14        Nishant   299                  4               5
14    15  __riddhi_213_   504                  2               7
15    16  Amrit Malviya   336                  2               4
16    17    darshimalde  3169                  3               3
17    18    Rohit Dutta   266                  3               3
18    19  fardinkamal62   263                  2               2
19    20         Shagun   100                  2               2
20    21         Vatsal  3408                  2               2

## Individual LeaderBoard

In [15]:
import re
# 'name' column is modified to give priority to alphabets rather than special characters
def order_string(text):
  text = text.upper()
  text = text[0] + re.sub(r'[^a-zA-Z]', '', text[1:])
  return(text)

### The individual leader board is ordered using the count of total statements and resaons given by a person in descending order, if this total count matches then they use alphabetical order prioritizing alphabets over special characters. 

In [16]:
# it is always a good option to copy existing data frame rather than modifying it which can leads to some errors
r1 = rigor.copy()
# count of total statements and reasons given by a person
r1['total'] = r1['total_statements'] + r1['total_reasons']
# modify name prioritizing alphabets over special characters
r1['modified_name'] = r1['name'].apply(order_string)

In [17]:
r2 = r1.sort_values(by=['total','modified_name'],ascending = [False,True], key=lambda x: [(not isinstance(val, str), val) for val in x])

In [18]:
# reset index and drop unwanted columns
r2.reset_index(drop = True, inplace = True)
r2 = r2.drop(['S No','total','modified_name'], axis=1)
r2 = r2.reset_index(drop=True).rename(columns={'name' : 'Name','uid' : 'UID','total_statements' : 'No. of Statements','total_reasons' : 'No. of Reasons'})
# set rank column based on order
r2.set_index([pd.Index(list(range(1,len(r2)+1)))],inplace=True)	
r2 = r2.reset_index(drop=False).rename(columns={'index': 'Rank'})	
r2																		

Rank           Name   UID  No. of Statements  No. of Reasons
0      1        Ronak 0   302                 16              19
1      2      Soumita M   280                 13              21
2      3     Subhangi 0    75                 13              16
3      4         Merwin   295                 13              12
4      5    Nitin Shane  1124                 12              11
5      6         Ayisha  3406                 11               8
6      7       imshawan    71                  9               9
7      8        sharath  3367                  8               8
8      9   Anuraj_Saini   342                  7               7
9     10       devmenkr   360                  7               7
10    11        Saurabh   271                  6               8
11    12         Palash    69                  6               7
12    13          raman   539                  5               5
13    14        Nishant   299                  4               5
14    15  __riddhi_213_   504                  2               7
15    16  Amrit Malviya   336                  2               4
16    17    darshimalde  3169                  3               3
17    18    Rohit Dutta   266                  3               3
18    19  fardinkamal62   263                  2               2
19    20         Shagun   100                  2               2
20    21         Vatsal  3408                  2               2

## Check if given output matches with computed output

In [19]:
# Check if the two DataFrames are equal
if r2.equals(Individual_leader_board):
    print("The two DataFrames are equal")
else:
    print("The two DataFrames are not equal")

The two DataFrames are equal


## TeamWise LeaderBoard

In [20]:
# copy user_ids input
u1 = User_Ids.copy()
# brandtech lab is written in two forms even though both indicating same team, so modify that to a single name
u1['Team Name'] = u1['Team Name'].replace('Brandtech Lab','BrandTech Lab')
# merge 2 dataframes based on 'name' column
user_team = pd.merge(u1, rigor, left_on='Name', right_on='name')
# group by team name
team = user_team.groupby('Team Name',as_index = False).mean()
# rename columns as given in output
team = team.rename(columns={'Team Name':'Thinking Teams Leaderboard','total_statements' : 'Average Statements', 'total_reasons' : 'Average Reasons'})
# drop unwanted columns
team = team.drop(['S No_x','User ID','S No_y','uid'], axis=1)
# sort by average statements made by each group and then by average reasons in descending order
team = team.sort_values(by=['Average Statements','Average Reasons'],ascending = [False,False])
# reset the index as team rank
team.set_index([pd.Index(list(range(1,len(team)+1)))],inplace=True)	
team = team.reset_index(drop=False).rename(columns={'index': 'Team Rank'})
team

Team Rank Thinking Teams Leaderboard  Average Statements  Average Reasons
0          1        Winning Culture Lab           12.666667        16.000000
1          2                   Tech Lab            9.000000        10.000000
2          3                     Mentor            7.333333         7.666667
3          4        Student Unicorn Lab            7.333333         6.333333
4          5                  Growpital            5.000000         5.666667
5          6              BrandTech Lab            4.500000         5.500000
6          7                    Kringle            4.500000         4.500000
7          8          Human Capital Lab            2.000000         4.500000

In [21]:
# Check if the two DataFrames are equal
if team.equals(team_leader_board):
    print("The two DataFrames are equal")
else:
    print("The two DataFrames are not equal")

The two DataFrames are not equal


### The above result shows the two dataframes are not equal this is because the average reasons of 'Tech Lab' doesn't match in both dataframes.
### According to the computed dataframe (team) the average reasons is the avergae of total reasons given by each person from the same team. 
### In this case imshawan, Ronak 0, Shagun belongs to Tech Lab team and they made 9, 19, 2 reasons respectively.
### Average_reasons = (9+19+2)/3 = 30/3 = 10
### But the given output contains 9.33 instead of 10, so this is the reason for difference in datframes.